File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from itertools import permutations
import math
def solution(numbers):
    answer = 0
    num_lst = []
    for i in range(1, len(numbers)+1):
        p = permutations(numbers, i)
        p = list(p)
        for idx in range(len(p)): #동일한게 나올 수 있음
            num_lst.append("".join(list(p[idx])))
    num_lst = list(map(int, num_lst))
    num_lst = list(set(num_lst))

    for num in num_lst:
        isPrime = True
        if num == 2:
            answer += 1
            continue
        if num == 1 or num == 0:
            isPrime = False
            continue
        for k in range(2, int(math.sqrt(num))+1):
            if num % k == 0:
                isPrime = False
                break
        if isPrime:
            answer += 1

    return answer

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;
class Solution {
    Set<Integer> set = new HashSet<>();
    public int solution(String numbers) {
        int answer = 0;
        boolean[] visited = new boolean[numbers.length()];
        
        dfs("", numbers, visited);
        return set.size();
       
    }
    private void dfs(String number, String numbers, boolean[] visited) {
        if (!number.isEmpty()) {
            int num = Integer.parseInt(number);
            if (isPrime(num)) {
                set.add(num);
            }
        } 
        for (int i = 0; i < numbers.length(); i++) {
            if (!visited[i]) {
                visited[i] = true;
                dfs(number + numbers.charAt(i), numbers, visited);
                visited[i] = false;
            }
        }
        
    }
    
    private boolean isPrime(Integer n) {
        
        if (n == 0 || n == 1) return false;
    
        for (int i = 2; i <= Math.sqrt(n); i++) {
            if (n % i == 0) return false;
        }
        return true;
    }
}